packages/amesos2/src/Amesos2_Tacho_decl.hpp

```diff
@@ -204,43 +222,42 @@ class TachoSolver : public SolverCore<Amesos2::TachoSolver, Matrix, Vector>
 
   // struct holds all data necessary to make a tacho factorization or solve call
   mutable struct TACHOData {
-    typename Tacho::Solver<tacho_type,SchedulerType> solver;
+    typename Tacho::Solver<tacho_type, scheduler_type> solver;
 
     // TODO: Implement the paramter options - confirm which we want and which have been implemented
     // int num_kokkos_threads;
     // int max_num_superblocks;
   } data_;
 
-  // The following Arrays are persisting storage arrays for A, X, and B
-  /// Stores the values of the nonzero entries for Tacho
-  Teuchos::Array<tacho_type> nzvals_;
-  /// Stores the location in \c Ai_ and Aval_ that starts row j
-  Teuchos::Array<ordinal_type> colind_;
-  /// Stores the row indices of the nonzero entries
-  Teuchos::Array<size_type> rowptr_;
-
-  // TODO: Decide handling for CUDA - how to fail
-#ifdef KOKKOS_ENABLE_OPENMP
-  typedef Kokkos::OpenMP DeviceSpaceType;
-#else
-  typedef Kokkos::Serial DeviceSpaceType;
-#endif
-  typedef typename Tacho::Solver<tacho_type,SchedulerType>::value_type_matrix
-    solve_array_t;
+  typedef typename Tacho::Solver<tacho_type, scheduler_type>::value_type_matrix
+    device_solve_array_t;
 
   // used as an internal workspace - possibly we can store this better in TACHOData
-  mutable solve_array_t workspace_;
+  mutable device_solve_array_t workspace_;
+
+  // x and b for solve - only allocated first time
+  mutable device_solve_array_t xValues_;
+  mutable device_solve_array_t bValues_;
+
+  // numeric is on device
+  device_value_type_array device_nzvals_view_;
+
+  // symbolic is done on host for Tacho so store these versions as well
+  host_size_type_array host_row_ptr_view_;
+  host_ordinal_type_array host_cols_view_;
 };                              // End class Tacho
 
 
 // Specialize solver_traits struct for Tacho
 template <>
 struct solver_traits<TachoSolver> {
 #ifdef HAVE_TEUCHOS_COMPLEX
-  typedef Meta::make_list4<float,
+  typedef Meta::make_list6<float,
```

Comment:
We have C++11 now; we can use `std::tuple` here.

Reply:
@kyungjoo-kim @ndellingwood @srajama1 If we did a change for this I expect we'd want to do it as a separate PR to cover all the solvers at once. We need the compile check if the type is in the list. make_listx sets this up with a recursive head + tail variable implementation. If we used tuple I expect we need some extra machinery to scan the types. I'm not sure yet what the cleanest way to scan the tuple types in C++11 is. I can investigate if you want.

Reply:
@MicheldeMessieres You don't need to address this comment now. We can come back later. Just make sure to keep it in the todo.